### PR TITLE
refs #245 add test to check connection leaking

### PR DESCRIPTION
**Pull request by kamichidu**
I add some tests to check connection leaking.
Now, go test must be failed, we can find connection leaking point by failing its test.

### CODE REVIEW

**Comment by kamichidu** [comment]
I will fix this problem later

File: tests/common_test.go [hunk 1]
```diff
@@ -41,6 +45,11 @@ func (s *BaseTestSuite) SetupSuite() {
 		panic(fmt.Sprintf("It was unable to connect to the DB.\n%s\n", err))
 	}
 
+	// set all connections will be closed immediately.
+	// this is required to check connections are leaked or not.
+	// because database/sql keep connection in the pool by default.
+	db.SetMaxIdleConns(0)
```
Comment (roobre) [inline tests/common_test.go]: You seem to enable this for all tests. Couldn't this be set just for testing idle conns instead of for the whole suite?

File: tests/common_test.go [hunk 2]
```diff
@@ -113,6 +131,24 @@ func (s *BaseTestSuite) resultOrError(res interface{}, err error) bool {
 	return true
 }
 
+func (s *BaseTestSuite) resultsOrError(res interface{}, err error) bool {
```
Comment (roobre) [inline tests/common_test.go]: What is this new function used for, exactly? Is it necessary?

File: tests/store_test.go [hunk 3]
```diff
@@ -111,6 +112,64 @@ func (s *StoreSuite) TestStoreFindOneReturnValues() {
 	s.resultOrError(doc, err)
 }
 
+func (s *StoreSuite) TestStoreFindAllReturnValues() {
```
Comment (roobre) [inline tests/store_test.go]: This seems to be unrelated to the PR. What does it test for?

File: tests/store_test.go [hunk 4]
```diff
@@ -111,6 +112,64 @@ func (s *StoreSuite) TestStoreFindOneReturnValues() {
...
+	docs, err = store.FindAll(NewStoreWithConstructFixtureQuery().Order(kallax.Asc(Schema.StoreWithConstructFixture.Foo)))
+	s.resultsOrError(docs, err)
+	s.NotPanics(func() {
+		s.Equal(2, len(docs))
+		s.Equal("bar", docs[0].Foo)
+		s.Equal("foo", docs[1].Foo)
+	})
+}
+
+func (s *StoreSuite) TestStoreCount() {
```
Comment (roobre) [inline tests/store_test.go]: Same as above

**Comment by roobre** [comment]
Also, if you intend to submit a fix, I'd prefer to merge it together with the test case.

Thank you in advance for the detailed report and efforts towards fixing this.

**Comment by kamichidu** [comment]
> I see more changes than I expected for just checking idle connections.

I wrote more tests unrelated this issue.
Because testing connection leaking is not only for current issue, I wanna write tests generated store methods for testing leaking.

> Also, if you intend to submit a fix, I'd prefer to merge it together with the test case.

Yep, I intend to submit a fix.
Maybe fixed on today or tomorrow.

**Comment by kamichidu** [comment]
And will reply your review ASAP, thanks for your review.

File: tests/common_test.go (hunk 1 above)
Comment (kamichidu) [reply tests/common_test.go]: It is okay be set just for testing idle conns, but I think about feature changes.
In the future, someone write some tests without setting idle conns checking, the test says okay with leaked idle conns.
Then I prefer to be set the whole test suite than to be set just for testing.

File: tests/common_test.go (hunk 2 above)
Comment (kamichidu) [reply tests/common_test.go]: It's not necessary at all.
This function is for writing a test for FindAll() with with the same way of FindOne() tests.
I will remove this function.

File: tests/store_test.go (hunk 3 above)
Comment (kamichidu) [reply tests/store_test.go]: Yes, this is unrelated.
Now, FindAll() is okay about leaking, but in the future?
I intend to detect leaking by test for some basic functions exposed by generated store.

File: tests/store_test.go (hunk 4 above)
Comment (kamichidu) [reply tests/store_test.go]: Same as above

**Comment by kamichidu** [comment]
Submitted a fix

**Comment by roobre** [comment]
Thanks for the detailed explanation and fixes.

I'm ok with merging this. However, I need to manually check and run some tests, as well as tweaking some spelling/grammar on comments and error messages. I'll merge into devel in a few days, when I return from vacation. I calculate this will reach master around next monday.

In the meantime, It will be helpful if you could squash the two commits into one.

Thanks again for the report+fix.

**Comment by kamichidu** [comment]
Thanks for your response.

> In the meantime, It will be helpful if you could squash the two commits into one.

Okay, I'll do that tomorrow.

Enjoy the vacation, thank you.

**Comment by kamichidu** [comment]
Squashed 2 into 1

**Comment by kamichidu** [comment]
what's up?

**Comment by roobre** [comment]
I've been addressing other issues, this is getting patched and merged soon, don't worry.

**Comment by e-sekito** [comment]
rebased

**Comment by L11R** [comment]
Why this super important PR still not merged? Our app built with kallax continues to exhaust all available DB connections cause the application to crash.

**Comment by L11R** [comment]
Before and After fix:
изображение
@roobre @erizocosmico please merge this ASAP.

**Comment by L11R** [comment]
@kamichidu what about community fork? Seems like @src-d abandoned it. Should you still use it in production?

**Comment by L11R** [comment]
@mcuadros, check it please.

**Comment by kamichidu** [comment]
> 
> 
> @kamichidu what about community fork? Seems like @src-d abandoned it. Do you still use it in production?

im not interest, sorry.
since i afraid kallax maintenance, i have developed my original orm like library.
then i and my team move to use my [goen](https://github.com/kamichidu/goen) instead of kallax.

**Comment by L11R** [comment]
@kamichidu got it, thanks for the project.

**Comment by L11R** [comment]
So seems like project is abandoned. We have to do private fork :\

**Comment by mcuadros** [comment]
Sorry, about the radio silence, the project is not abandoned, but looks like the external maintainer is not maintaining it anymore. 

I will try to review and merge the other PRs ASAP. Thanks!

**Comment by savely-krasovsky** [comment]
@mcuadros thanks a lot!
